src/core/libMeta.ml

```diff
@@ -0,0 +1,111 @@
+
+(** [fresh_box p ?name a n] is the boxed counterpart of [fresh_meta]. It is
+   only useful in the rare cases where the type of a metavariable contains a
+   free term variable environment. This should only happens when scoping the
+   rewriting rules, use this function with care.  The metavariable is created
+   immediately with a dummy type, and the type becomes valid at unboxing. The
+   boxed metavariable should be unboxed at most once, otherwise its type may
+   be rendered invalid in some contexts. *)
+let fresh_box: problem -> ?name:string -> tbox -> int -> meta Bindlib.box =
+  fun p ?name a n ->
+  let m = fresh p ?name mk_Kind n in
+  Bindlib.box_apply (fun a -> m.meta_type := a; m) a
+
+(** [set p m v] sets the metavariable [m] of [p] to [v]. WARNING: No specific
+   check is performed, so this function may lead to cyclic terms. To use with
+   care. *)
+let set : problem -> meta -> tmbinder -> unit = fun p m v ->
+  m.meta_type := mk_Kind; (* to save memory *) m.meta_value := Some v;
+  p.metas <- MetaSet.remove m p.metas
+
+(** [name m] returns a string representation of [m]. *)
+let name : meta -> string = fun m ->
+  match m.meta_name with
+  | Some n -> n
+  | None -> string_of_int m.meta_key
+
+(** [of_name p n] returns [Some m] if [m] is an element of the set of metas of
+   [p] with name [n], and [None] otherwise. *)
+let of_name : string -> problem -> meta option = fun n p ->
+  let exception Found of meta in
+  let f m = if m.meta_name = Some n then raise (Found m) in
+  try MetaSet.iter f p.metas; None with Found m -> Some m
+
+(** [make p ctx a] creates a fresh metavariable term of type [a] in the
+   context [ctx], and adds it to [p]. *)
+let make : problem -> ctxt -> term -> term = fun p ctx a ->
+  let a, k = Ctxt.to_prod ctx a in
+  let m = fresh p a k in
+  let get_var (x,_,_) = mk_Vari x in
+  mk_Meta(m, Array.of_list (List.rev_map get_var ctx))
+
+(** [make_codomain p ctx a] creates a fresh metavariable term of type [Type]
+   in the context [ctx] extended with a fresh variable of type [a], and
+   updates [p] with generated metavariables. *)
+let make_codomain : problem -> ctxt -> term -> tbinder = fun p ctx a ->
+  let x = new_tvar "x" in
+  let b = make p ((x, a, None) :: ctx) mk_Type in
+  Bindlib.unbox (Bindlib.bind_var x (lift b))
+(* Possible improvement: avoid lift by defining a function _Meta.make
+   returning a tbox. *)
+
+(** [iter b f c t] applies the function [f] to every metavariable of [t] and,
+   if [x] is a variable of [t] mapped to [v] in the context [c], then to every
+   metavariable of [v], and to the type of every metavariable recursively if
+   [b] is true. *)
+let iter : bool -> (meta -> unit) -> ctxt -> term -> unit = fun b f c ->
+  (* Convert the context into a map. *)
+  let vm =
+    let f vm (x,_,v) =
+      match v with
+      | Some v -> VarMap.add x v vm
+      | None -> vm
+    in
+    Stdlib.ref (List.fold_left f VarMap.empty c)
+  in
+  (*TODO: make iter tail recursive. *)
```

Comment:
Why do you want to make it tail recursive?

Reply:
Well, to make it more efficient.

Reply:
I'm not sure it's always more efficient: in some cases it's better to keep not tail recursive functions when the data structures are small. But that's not important for now.

Reply:
I don't know either but this is something to test. This could be a way also to handle the different kinds of checking problems in a single function. We'll see this later.